pkg/apis/core/validation/validation.go

```diff
@@ -1443,25 +1443,32 @@ func validateStorageOSPersistentVolumeSource(storageos *core.StorageOSPersistent
 	return allErrs
 }
 
-func validateCSIPersistentVolumeSource(csi *core.CSIPersistentVolumeSource, fldPath *field.Path) field.ErrorList {
+func ValidateCSIDriverName(driverName string, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIPersistentVolume) {
-		allErrs = append(allErrs, field.Forbidden(fldPath, "CSIPersistentVolume disabled by feature-gate"))
+	if len(driverName) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
 	}
 
-	if len(csi.Driver) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("driver"), ""))
+	if len(driverName) > 63 {
+		allErrs = append(allErrs, field.TooLong(fldPath, driverName, 63))
 	}
 
-	if len(csi.Driver) > 63 {
-		allErrs = append(allErrs, field.TooLong(fldPath.Child("driver"), csi.Driver, 63))
+	if !csiDriverNameRexp.MatchString(driverName) {
```

Comment:
keep in mind that we cannot change the v1beta1 API. additional validation like this we want to do needs to be done only on creation, and only when the API request is not via the v1beta1 API.

That info is available in the context passed to volumeAttachmentStrategy#Validate, the same way this PR is doing different behavior in PrepareForCreate for v1beta1

Reply:
@liggitt, fixed. I squashed two validation commits into one and introduced `ValidateVolumeAttachmentV1` that's called only when v1 instance is created. I hope I got it right.

Reply:
If I understand it correctly, this is just a temporary measure, right? VolumeAttachments are re-created during each Kubernetes update (nodes get drained -> all volumes are detached -> all VolumeAttachments are deleted). At some point Kubernetes will create v1 objects only, so we can merge the validation functions.

Or shall we support 3rd party external attachers that are going to use v1beta1 API forever?

Reply:
once we have v1, we can deprecate v1beta1 and stop serving it after the deprecation period. until we have the ability to sweep and remove existing invalid objects, we will have to tolerate data that could have been persisted via v1beta1